package.json

```diff
@@ -21,6 +21,17 @@
   "jest": {
     "testTimeout": 60000,
     "clearMocks": true,
+    "collectCoverageFrom": [
+      "**/*.ts"
+    ],
+    "coverageThreshold": {
+      "global": {
+        "branches": 98,
+        "functions": 98,
+        "lines": 98,
+        "statements": 98
+      }
+    },
```

Comment:
> The last couple of percent are usually unreasonably difficult to obtain - rarely encountered edge case errors which are hard to simulate and take unreasonable amount of time to write tests for.

Yeah, there are a couple of cases I haven't been able to write tests for for this reason. For example, one line that only is run in the browser and never triggered in tests.

A workaround would be to (sparingly) add tags to disable coverage reporting for specific lines. The above case might be the only one where it's absolutely needed. Considering that, I believe we can attain and maintain 100% coverage.

> People tend to strive for those 100% and once they reach them they get a false sense of security

Interesting psychological point. I feel like having all code covered is better than having some statements/branches not tested at all. It's true that 100% does not really mean 100% and there's a lot of combinations that aren't tested (true full coverage not being realistically possible.) We just have to keep that in mind when writing tests.

Do you have any sources I could read up further on regarding this point? It reminds me of the "helmets make you feel safer and thus more likely to get in an accident" argument. It would be nice if someone has collected actual data on bugs caught/missed with/without full coverage.

> If we set it to 90% that would be just as good, as we won't be able to make meaningful code changes without adding tests but we'll be able to quickly fix something if need be.

Hmm I'm not sure we want to be pushing quick untested changes anyway. Having coverage reporting, and strict rules about every line being tested, earlier would have prevented a few bugs from making it to production.

Reply:
> "helmets make you feel safer and thus more likely to get in an accident"
Ha ha, yeah, it is something like that!

> Do you have any sources I could read up further on regarding this point? 
No, I'm afraid I don't - I might try digging around though. My only source is my experience but that's always biased, based on the people I worked with in the past.

Last point: setting the minimum coverage level to 90% does not stop you in any way from adding all the tests you want and reaching 100%.

Anyway, this was only a suggestion. If you feel 100% is appropriate and reasonably achievable then let's go for that.